### DISCUSSION TITLE
Build Failed Command "npm run build" exited with 1

### DISCUSSION THREAD

**Discussion by bchang0999** [vercel/next.js#41114] [Help] [answered]
Hey guys im trying to deploy my project. 
I run Vercel Build through my terminal it comes back with no errors but when run vercel deploy I keep getting 
```Build Failed Command "npm run build" exited with 1```

Any suggestions or fixes would be greatly apreciated!
I see it says I need a projectId but my .env.local contains my id 
```NEXT_PUBLIC_SANITY_DATASET=production
NEXT_PUBLIC_SANITY_PROJECT_ID=8a64h111
NEXT_PUBLIC_BASE_URL=http://localhost:3000/```
 
I'm really stumped here.

```23:27:55.917 | Error: Configuration must contain `projectId`
23:27:55.917 | at exports.initConfig (/vercel/path0/node_modules/@sanity/client/lib/config.js:42:11)
23:27:55.918 | at SanityClient.config (/vercel/path0/node_modules/@sanity/client/lib/sanityClient.js:89:25)
23:27:55.918 | at new SanityClient (/vercel/path0/node_modules/@sanity/client/lib/sanityClient.js:53:8)
23:27:55.918 | at SanityClient (/vercel/path0/node_modules/@sanity/client/lib/sanityClient.js:50:12)
23:27:55.918 | at exports.createClient (/vercel/path0/node_modules/next-sanity/dist/next-sanity.cjs.production.min.js:1:1023)
23:27:55.918 | at Object.5149 (/vercel/path0/.next/server/pages/index.js:1055:75)
23:27:55.919 | at __webpack_require__ (/vercel/path0/.next/server/webpack-runtime.js:25:42)
23:27:55.919 | at /vercel/path0/.next/server/pages/index.js:20:65
23:27:55.919 | at Function.__webpack_require__.a (/vercel/path0/.next/server/webpack-runtime.js:89:13)
23:27:55.919 | at Object.2265 (/vercel/path0/.next/server/pages/index.js:11:21)
23:27:55.920 |  
23:27:55.920 | > Build error occurred
23:27:55.922 | Error: Failed to collect page data for /
23:27:55.922 | at /vercel/path0/node_modules/next/dist/build/utils.js:916:15 {
23:27:55.922 | type: 'Error'
23:27:55.923 | }
23:27:55.978 | Error: Command "npm run build" exited with 1```

**Comment by icyJoseph** [comment, marked as answer]
Are you including your `.env.local` to the repository (you shouldn't anyway, so don't do it)? Or are you injecting it in the `Vercel` dashboard and doing a `vercel pull <environment>` kind of thing?

**Edit:** I found your repo and you CORRECTLY skip including `.env.local` to the repository, but you have to do it in the `Vercel` dashboard. 

Normally you need to go on the `Vercel` dashboard and define the env variables there, for preview and production environments. 

I cloned your project and ran `npm run build` without env vars and I got the exact same error, so it must be that you haven't defined them in the `Vercel` dashboard.

I do believe that the `NEXT_PUBLIC_BASE_URL` will probably give you trouble later, but we'll see.

**Reply by icyJoseph** [reply]
@bchang0999 Did you use this UI to load your variables?

<img width="1012" alt="Screenshot 2022-10-03 at 19 35 13" src="https://user-images.githubusercontent.com/21013447/193642161-157f0186-9197-4f60-bbdf-be14083ab792.png">

You don't have to define a file manually, they'll make the variables available to the application, at build and runtime.

As for the new error you are seeing, it happens because you are trying to access `getPageInfo` via `/api` routes, which do not exist at build time... `/api/routes` exist only when the server is running. You have to connect to the whatever `getPageInfo` uses directly at build time, hopefully you can do this from `getStaticProps`, or `getServerSideProps`. 

I'll take a look around your repo to see where you make the call and what alternative you can take.

**Reply by bchang0999** [reply]
Yes thats the exact UI I used
![Screenshot (15)](https://user-images.githubusercontent.com/101777205/193643369-087c530a-9175-4a20-b5b8-a5ddcb17a50b.png)
This is how I filled it out for my api get requests 
![Screenshot (18)](https://user-images.githubusercontent.com/101777205/193643670-b339096b-6e50-46dd-b7cf-0b231cdc903e.png)

and these are the fetch request 
![Screenshot (19)](https://user-images.githubusercontent.com/101777205/193643721-080b74e1-34df-4aab-a6db-ad32eca4a4a7.png)

They are all the same for the most part some are just a little different depending on what they are trying to pull, but thank you so much for getting back to me again

**Reply by icyJoseph** [reply]
Alright, regarding `getPageInfo`, the way you have it is already good architecture, but there's a piece that should be changed:

```ts
// utils/fetchPageInfo.ts
import { PageInfo } from "../typings";
import { groq } from "next-sanity";
import { sanityClient } from "../sanity";

const pageInfoQuery = groq`
  *[_type == "pageInfo"][0]
`;

export const fetchPageInfo = async () => {
  try {
    const pageInfo: PageInfo = await sanityClient.fetch(pageInfoQuery);
    // here you could even try to validate the shape of pageInfo, or some other business logic, and throw is necessary
    return pageInfo;
  } catch (reason) {
    if (reason instanceof Error) {
      // do something with the error
      // or just
      throw reason;
    }
    // doesn't always work to do this, but hopefully reason
    // is serializable as a string
    throw new Error(`Unexpected exception ${reason}`);
  }
};
```

and then in API routes:

```ts
import type { NextApiRequest, NextApiResponse } from "next";
import { PageInfo } from "../../typings";
import { fetchPageInfo } from "../../utils/fetchPageInfo";

type Data = {
  pageInfo: PageInfo;
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {
  try {
    const pageInfo = await fetchPageInfo();

    return res.status(200).json({ pageInfo });
  } catch (error) {
    return res.status(500).end();
  }
}
```

Or as you already have in the `pages/index` getStaticProps function.

**Reply by bchang0999** [reply]
It worked it got past the initial error and moved on to ```FetchError: request to http://localhost:3000/api/getExperience failed, reason: connect ECONNREFUSED 127.0.0.1:3000```   I'm assuming I would just have to change it for all of them

**Reply by icyJoseph** [reply]
Yeah, that's about right, it is architecture forced by the framework, if one could frame it that way, but you had the right idea!

You can also just run `next build` locally, and it should print the same errors.

Hope you get it running! Don't forget to mark as answered!

**Reply by bchang0999** [reply]
I'm sorry could you help me re-write the experience section, the plurals are throwing me off and I wont forget to

**Reply by icyJoseph** [reply]
Did you manage to the experience section working?

**Reply by bchang0999** [reply]
I'm having trouble with plurals and relationship of it

**Reply by icyJoseph** [reply]
Ah, sorry I haven't played with `groq` enough to be able to give advice.

**Reply by bchang0999** [reply]
Thank you for all the help though!

**Reply by icyJoseph** [reply]
Maybe close this one, since it's answered anyway, and find the `groq` forums to get help with that.

**Comment by bchang0999** [comment]
On vercel when adding env variables, the name I put was NEXT_PUBLIC_SANITY_PROJECT_ID (names couldnt have symbols in them) and then I added in value 
abcdefgh (whatever the id is)
I did the same for NEXT_PUBLIC_SANITY_DATASET
NEXT_PUBLIC_SANITY_PROJECT_ID
NEXT_PUBLIC_BASE_URL

Was this the correct way to add it?
Also a new error occurs 

Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
--
13:25:04.823 | FetchError: request to http://localhost:3000/api/getPageInfo failed, reason: connect ECONNREFUSED 127.0.0.1:3000
13:25:04.824 | at ClientRequest.<anonymous> (/vercel/path0/node_modules/next/dist/compiled/node-fetch/index.js:1:65756)
13:25:04.824 | at ClientRequest.emit (node:events:527:28)
13:25:04.825 | at Socket.socketErrorListener (node:_http_client:454:9)
13:25:04.825 | at Socket.emit (node:events:527:28)
13:25:04.825 | at emitErrorNT (node:internal/streams/destroy:157:8)
13:25:04.826 | at emitErrorCloseNT (node:internal/streams/destroy:122:3)
13:25:04.826 | at processTicksAndRejections (node:internal/process/task_queues:83:21)
13:25:04.827 |  
13:25:04.827 | > Build error occurred
13:25:04.828 | info  - Generating static pages (3/3)
13:25:04.828 | Error: Export encountered errors on following paths:
13:25:04.829 | /
13:25:04.829 | at /vercel/path0/node_modules/next/dist/export/index.js:404:19
13:25:04.829 | at runMicrotasks (<anonymous>)
13:25:04.830 | at processTicksAndRejections (node:internal/process/task_queues:96:5)
13:25:04.830 | at async Span.traceAsyncFn (/vercel/path0/node_modules/next/dist/trace/trace.js:79:20)
13:25:04.830 | at async /vercel/path0/node_modules/next/dist/build/index.js:1229:21
13:25:04.831 | at async Span.traceAsyncFn (/vercel/path0/node_modules/next/dist/trace/trace.js:79:20)
13:25:04.831 | at async /vercel/path0/node_modules/next/dist/build/index.js:1091:17
13:25:04.832 | at async Span.traceAsyncFn (/vercel/path0/node_modules/next/dist/trace/trace.js:79:20)
13:25:04.832 | at async Object.build [as default] (/vercel/path0/node_modules/next/dist/build/index.js:65:29)
13:25:04.891 | Error: Command "npm run build" exited with 1

**Comment by pranjalkv** [comment]
You can also put your .env.local data in next.config.js (now you can delete .env.local)
`module.exports = {
  env: {
    API_KEY: 'xyz',
  },
}
`
and get that value using process.env.API_KEY
